presto-hive/src/main/java/io/prestosql/plugin/hive/ReaderProjectionsAdapter.java

```diff
@@ -0,0 +1,245 @@
+            this.type = type;
+        }
+
+        @Override
+        public Block load()
+        {
+            checkState(!loaded, "Already loaded");
+            Block loadedBlock = loadInternalBlock(dereferenceSequence, inputBlock);
+            inputBlock = null;
+            loaded = true;
+            return loadedBlock;
+        }
+
+        /**
+         * Applies dereference operations on the input block to extract the required internal block. If the input block is lazy
+         * in a nested manner, this implementation avoids loading the entire input block.
+         */
+        private Block loadInternalBlock(List<Integer> dereferences, Block parentBlock)
+        {
+            if (dereferences.size() == 0) {
+                return parentBlock.getLoadedBlock();
+            }
+
+            ColumnarRow columnarRow = toColumnarRow(parentBlock);
+
+            int dereferenceIndex = dereferences.get(0);
+            List<Integer> remainingDereferences = dereferences.subList(1, dereferences.size());
+
+            Block fieldBlock = columnarRow.getField(dereferenceIndex);
+            Block loadedInternalBlock = loadInternalBlock(remainingDereferences, fieldBlock);
+
+            // Field blocks provided by ColumnarRow can have a smaller position count, because they do not store nulls.
+            // The following step adds null elements (when required) to the loaded block.
+            return adaptNulls(columnarRow, loadedInternalBlock);
+        }
+
+        private Block adaptNulls(ColumnarRow columnarRow, Block loadedInternalBlock)
+        {
+            BlockBuilder newlyCreatedBlock = null;
+            int fieldBlockPosition = 0;
+
+            for (int i = 0; i < columnarRow.getPositionCount(); i++) {
+                boolean isRowNull = columnarRow.isNull(i);
+
+                if (isRowNull) {
+                    // A new block is only created when a null is encountered for the first time.
+                    if (newlyCreatedBlock == null) {
+                        newlyCreatedBlock = type.createBlockBuilder(null, columnarRow.getPositionCount());
+
+                        // Copy over all elements encountered so far to the new block
+                        for (int j = 0; j < i; j++) {
+                            type.appendTo(loadedInternalBlock, j, newlyCreatedBlock);
+                        }
+                    }
+                    newlyCreatedBlock.appendNull();
+                }
+                else {
+                    if (newlyCreatedBlock != null) {
+                        type.appendTo(loadedInternalBlock, fieldBlockPosition, newlyCreatedBlock);
+                    }
+                    fieldBlockPosition++;
+                }
+            }
+
+            if (newlyCreatedBlock == null) {
+                // If there was no need to create a null, return the original block
+                return loadedInternalBlock;
+            }
+
+            return newlyCreatedBlock.build();
+        }
```

Comment:
Hmm... this is not ideal. There will be a noticeable performance hit due to the extra copying and assembly. What we need is a way to assemble a block for this column with the "null vector" from the parent row block and block for the field. @dain, any thoughts on this?

Reply:
@martint agreed. I believe we'll need to assemble a block by reindexing, considering nulls. similar to https://github.com/prestosql/presto/blob/ed99fb27e35605da80fda28f1b5c10424e3bc53a/presto-spi/src/main/java/io/prestosql/spi/block/RowBlock.java#L48

I don't know if there is any existing Block implementation that provides this. `DictionaryBlock` won't work, because its implementation does not allow representing nulls through negative indexes. If we modify that behavior, it'd probably require changing accessors of `DictionaryBlock`. Or we could create a new block implementation altogether for this purpose.

Reply:
@martint applying dereference projection in Hive connector means that we're getting rid of this projection in the engine (likely in SFP operator). We do have the same issue of copying data there as well, right? How do we resolve it right now in the engine?

Reply:
Yes, you're right. It loops over the values, gets the field for each row element and appends to a new block. Ok, this is probably fine for now, but there's an opportunity to improve performance later. Let's make a note of it in a TODO.

Reply:
I've added a TODO for `adaptNulls` method.